src/oscar/apps/wishlists/views.py

```diff
@@ -0,0 +1,48 @@
+from django.conf import settings
+from django.contrib import messages
+from django.contrib.auth import logout
+from django.core.paginator import Paginator
+from django.http import Http404
+from django.shortcuts import get_object_or_404, redirect
+from django.utils.translation import gettext_lazy as _
+from django.views.generic import DetailView
+
+from oscar.core.loading import get_model
+
+WishList = get_model('wishlists', 'WishList')
+
+
+class WishListView(DetailView):
+    template_name = 'oscar/wishlists/wishlist_detail.html'
+    model = WishList
+    context_object_name = 'wishlist'
+
+    def dispatch(self, request, *args, **kwargs):
+        wishlist = self.get_object()
+
+        if wishlist.is_allowed_to_see(request.user):
+            return super().dispatch(request, *args, **kwargs)
+        elif wishlist.visibility == WishList.PRIVATE:
+            raise Http404
+        else:
+            messages.info(request, _("Please login with the email address that is shared to this wish list."))
```

Comment:
I don't think this message is a clear way of handling not allowed to see the message. I think you should only redirect to login if the user is anonymoes, and simply state "You are not allowed to see this wishlist", 403 if sombody is logged in but not allowed to view the wishlist.